src/grid/grid_task_list.c

```diff
@@ -33,60 +41,119 @@ void grid_create_task_list(
+            "Please re-compile with -D__GRID_CUDA.");
+    abort();
+    break;
+#endif
+  case GRID_BACKEND_CPU: {
+    if (!(*task_list)->cpu) {
+      (*task_list)->cpu = create_grid_context_cpu(
+          ntasks, nlevels, natoms, nkinds, nblocks, buffer_size, block_offsets,
+          atom_positions, atom_kinds, basis_sets, level_list, iatom_list,
+          jatom_list, iset_list, jset_list, ipgf_list, jpgf_list,
+          border_mask_list, block_num_list, radius_list, rab_list,
+          blocks_buffer);
+    } else {
+      update_grid_context_cpu(
+          ntasks, nlevels, natoms, nkinds, nblocks, buffer_size, block_offsets,
+          atom_positions, atom_kinds, basis_sets, level_list, iatom_list,
+          jatom_list, iset_list, jset_list, ipgf_list, jpgf_list,
+          border_mask_list, block_num_list, radius_list, rab_list,
+          blocks_buffer, (*task_list)->cpu);
+    }
+    (*task_list)->backend = GRID_BACKEND_CPU;
+  } break;
 #ifdef __GRID_CUDA
+  case GRID_BACKEND_GPU:
     grid_gpu_create_task_list(
         ntasks, nlevels, natoms, nkinds, nblocks, buffer_size, block_offsets,
         atom_positions, atom_kinds, basis_sets, level_list, iatom_list,
         jatom_list, iset_list, jset_list, ipgf_list, jpgf_list,
         border_mask_list, block_num_list, radius_list, rab_list, blocks_buffer,
         &(*task_list)->gpu);
-#else
-    fprintf(stderr, "Error: The GPU grid backend is not available. "
-                    "Please re-compile with -D__GRID_CUDA.");
-    abort();
-#endif // __GRID_CUDA
-  } else if ((*task_list)->backend == GRID_BACKEND_REF) {
+    (*task_list)->backend = GRID_BACKEND_GPU;
+    break;
+  case GRID_BACKEND_HYBRID: {
+    if (!(*task_list)->hybrid) {
+      (*task_list)->hybrid = create_grid_context_cpu(
+          ntasks, nlevels, natoms, nkinds, nblocks, buffer_size, block_offsets,
+          atom_positions, atom_kinds, basis_sets, level_list, iatom_list,
+          jatom_list, iset_list, jset_list, ipgf_list, jpgf_list,
+          border_mask_list, block_num_list, radius_list, rab_list,
+          blocks_buffer);
+    } else {
+      update_grid_context_cpu(
+          ntasks, nlevels, natoms, nkinds, nblocks, buffer_size, block_offsets,
+          atom_positions, atom_kinds, basis_sets, level_list, iatom_list,
+          jatom_list, iset_list, jset_list, ipgf_list, jpgf_list,
+          border_mask_list, block_num_list, radius_list, rab_list,
+          blocks_buffer, (*task_list)->hybrid);
+    }
+
+    /* does not allocate anything on the GPU. */
+    /* allocation only occurs when the collocate (or integrate) function is
+     * called. Resources are released before exiting the function */
+
+    /* I do *not* store the address of config.device_id */
+    initialize_grid_context_on_gpu(
+        (*task_list)->hybrid, 1 /* number of devices */, &config.device_id);
+    update_queue_length((*task_list)->hybrid, config.queue_length);
+    (*task_list)->backend = GRID_BACKEND_HYBRID;
```

Comment:
This is a lot of code that could be moved into your backends. Maybe you can follow the existing scheme and introduce `grid_hybrid_create_task_list` and  `grid_cpu_create_task_list` routines? These routines could also contain the calls to `apply_cutoff`.

Reply:
I am afraid this will stay there. What is the point here. Unhappy with with the fact the name of function contains the functionality first and then on it applies to. I am not changing this. 

Also why should i remove the fact that I update a structure inside a function that creates it. Updating and creating are two different things that why I did it.

Reply:
My point is that this logic is an implementation detail of your backends that does not belong into this high level file.

While I really wish you'd just follow the scheme that I have laid out, I'm willing to merge this nevertheless as it does not inflict the other backends.